starter/src/main/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2Collector.java

```diff
@@ -42,8 +42,19 @@
 public class PSD2Collector extends X509Collector {
 
     @Builder(builderMethodName = "psd2Builder")
-    public PSD2Collector(String collectorName, UsernameCollector usernameCollector, AuthoritiesCollector authoritiesCollector, CertificateHeaderFormat collectFromHeader, String headerName) {
-        super(collectorName, usernameCollector, certificatesChain -> {
+    public PSD2Collector(String collectorName, PSD2UsernameCollector usernameCollector,
+                         AuthoritiesCollector authoritiesCollector, CertificateHeaderFormat collectFromHeader,
+                         String headerName) {
+        super(collectorName, certificatesChain -> {
```

Comment:
The lambdas here make it difficult to see where the call to the super constructor starts and ends, and which parameters are being provided.

Instead we could extract them into methods and have:

```
    super(collectorName,
              getUsernameCollector(usernameCollector),
              getAuthoritiesCollector(authoritiesCollector),
              collectFromHeader, headerName);
```

IntelliJ will give you the option to extract them into methods, but here are they are for convenience:

```
    private static UsernameCollector getUsernameCollector(PSD2UsernameCollector usernameCollector) {
        return certificatesChain -> {
            try {
                Psd2CertInfo psd2CertInfo = new Psd2CertInfo(certificatesChain);
                return usernameCollector.getUserName(certificatesChain, psd2CertInfo);
            } catch (InvalidPsd2EidasCertificate | InvalidEidasCertType invalidPsd2EidasCertificate) {
                log.warn("Certificate found couldn't be parsed as a PSD2 certificate. Certificate will be ignored",
                        invalidPsd2EidasCertificate);
                return null;
            }
        };
    }

    private static X509Collector.AuthoritiesCollector getAuthoritiesCollector(AuthoritiesCollector authoritiesCollector) {
        return certificatesChain -> {
            Set<GrantedAuthority> authorities = new HashSet<>();
            try {
                Psd2CertInfo psd2CertInfo = new Psd2CertInfo(certificatesChain);
                log.trace("Verify if certificate is a QWAC");
                if (psd2CertInfo.isPsd2Cert()
                        && psd2CertInfo.getEidasCertType().isPresent()
                        && psd2CertInfo.getEidasCertType().get().equals(EidasCertType.WEB)) {

                    //Map PSD2 roles
                    Optional<Psd2QcStatement> psd2QcStatementOpt = psd2CertInfo.getPsd2QCStatement();
                    if (psd2QcStatementOpt.isPresent()) {
                        Psd2QcStatement psd2QcStatement = psd2QcStatementOpt.get();
                        log.trace("Founds PSD2 QC Statement {}", psd2QcStatement);
                        authorities.addAll(authoritiesCollector.getAuthorities(certificatesChain, psd2CertInfo, psd2QcStatement.getRoles()));
                    } else {
                        log.trace("No PSD2 QC Statement found");
                        authorities.addAll(authoritiesCollector.getAuthorities(certificatesChain, psd2CertInfo, null));
                    }
                } else {
                    if (log.isTraceEnabled()) {
                        if (!psd2CertInfo.isPsd2Cert()) {
                            log.trace("Not a PSD2 cert");
                        } else if (psd2CertInfo.getEidasCertType().isEmpty()) {
                            log.trace("Is a PSD2 certs but no EIDAS cert type");
                        } else if (psd2CertInfo.getEidasCertType().isEmpty()) {
                            log.trace("Is a PSD2 certs with EIDAS cert type {} but it's not a QWAC", psd2CertInfo.getEidasCertType().get());
                        }
                    }
                    authorities.addAll(authoritiesCollector.getAuthorities(certificatesChain, null, null));
                }
            } catch (InvalidPsd2EidasCertificate | InvalidEidasCertType invalidPsd2EidasCertificate) {
                log.warn("Certificate founds couldn't be parsed as a PSD2 certificate. Will ignore the certificate", invalidPsd2EidasCertificate);
            }

            return authorities;
        };
    }
```

Reply:
Thanks Matt

Reply:
Fixed